### DISCUSSION TITLE
<ERROR> Problem connecting to /icubSim/left_leg

### DISCUSSION THREAD

**Discussion by jimingda** [robotology/community#646] [Q&A] [answered]
Hi,

I delete the `--config no_legs.xml` parameter in yarprobotinterface  on yarpmanager gui, but still cann't  connect to legs as the error in yarprobotinterface.log below. Shall i modify config somewhere else?

Thanks in advance.
 
```
09:22 # cat ~/log_db61dac93e4e_yarprobotinterface_89619 
 4.349416 <DEBUG> Reading file /usr/local/share/iCub/contexts/gazeboCartesianControl/./all.xml
 4.369697 <DEBUG> yarprobotinterface: using xml parser for DTD v3.x
 4.369741 <DEBUG> Preprocessor complete in:  0.00080204 s
 4.369755 <WARNING> *************************************************************************************
* yarprobotinterface 'portprefix' parameter does not follow convention,  *
* it MUST start with a leading '/' since it is used as the full prefix port name    *
*     name:    full port prefix name with leading '/',  e.g.  /robotName      *
* A temporary automatic fix will be done for you, but please fix your config file   *
*************************************************************************************
 4.411170 <INFO> startup phase starting...
 4.421235 <DEBUG> Parameters are (device remote_controlboard) (id torso) (local "/gazeboCartesianControl/torso") (remote "/icubSim/torso") (robotName icubSim) (writeStrict on)
 4.651077 <INFO> Created device <remote_controlboard>. See C++ class RemoteControlBoard for documentation.
 4.651137 <DEBUG> Parameters are (device remote_controlboard) (id left_arm) (local "/gazeboCartesianControl/left_arm") (remote "/icubSim/left_arm") (robotName icubSim) (writeStrict on)
 4.917796 <INFO> Created device <remote_controlboard>. See C++ class RemoteControlBoard for documentation.
 4.917899 <INFO> Opening device right_arm with parameters [("robotName" = "icubSim"), ("remote" = "/icubSim/right_arm"), ("local" = "/gazeboCartesianControl/right_arm"), ("writeStrict" = "on")]
 4.917926 <DEBUG> Parameters are (device remote_controlboard) (id right_arm) (local "/gazeboCartesianControl/right_arm") (remote "/icubSim/right_arm") (robotName icubSim) (writeStrict on)
 4.917964 <INFO> RemoteControlBoard is ENABLING the writeStrict option for all commands
 5.160272 <INFO> Created device <remote_controlboard>. See C++ class RemoteControlBoard for documentation.
 5.160326 <INFO> Opening device left_leg with parameters [("robotName" = "icubSim"), ("remote" = "/icubSim/left_leg"), ("local" = "/gazeboCartesianControl/left_leg"), ("writeStrict" = "on")]
 5.160342 <DEBUG> Parameters are (device remote_controlboard) (id left_leg) (local "/gazeboCartesianControl/left_leg") (remote "/icubSim/left_leg") (robotName icubSim) (writeStrict on)
 5.160364 <INFO> RemoteControlBoard is ENABLING the writeStrict option for all commands
 5.221792 <ERROR> Problem connecting to /icubSim/left_leg/rpc:i, is the remote device available?
 5.231911 <ERROR> Problem connecting to /icubSim/left_leg/command:i, is the remote device available?
 5.231976 <ERROR> Problem connecting to /icubSim/left_leg/stateExt:o, is the remote device available?
 5.231992 <ERROR> Driver <remote_controlboard> was found but could not open

```

**Comment by pattacini** [comment, marked as answer]
Hi @jimingda 

If you're talking about `tutorial_cartesian-interface`, then I think that the referenced [robot model][1] isn't shipped with controllable legs. Not sure though.
cc @xEnVrE @traversaro.

Alternatively, you may consider relying on a different model, like [iCubGazeboV2.7][2].

[1]: https://github.com/vvv-school/tutorial_cartesian-interface/blob/master/gazebo/worlds/tutorial_cartesian-interface.sdf#L21
[2]: https://github.com/robotology/icub-models/tree/master/iCub/robots/iCubGazeboV2_7

**Reply by jimingda** [reply]
Thanks@pattacini
Is there any method to integrate the V2.7 model. I directly replaced the iCubGazeboV2_5_visuomanip with iCubGazeboV2_7, and failed to launch the system apps like yarprobotinterface and iKinCartesianSolver.

**Reply by pattacini** [reply]
Hi @jimingda 

I've just tried it out and it worked, both with `iCubGazeboV2_7` and `iCubGazeboV2_5`.

Make sure that:
- You modify this file: https://github.com/vvv-school/tutorial_cartesian-interface/blob/master/gazebo/worlds/tutorial_cartesian-interface.sdf.
- You build the project from scratch soon afterward.

Also, probably you'll have to switch to the use of `iCubGazeboV2_5` as you've been using an old docker image, since `iCubGazeboV2_7` has been fixed only recently.

**Reply by jimingda** [reply]
Thanks and iCubGazeboV2_5 works well. Another question is that how does the three joints of hip below correspond to roll pitch and yaw angle?
![2023-08-01_17-33](https://github.com/robotology/community/assets/7019692/edaa1a5e-ee15-46a0-a403-10d3aa0379d7)

**Reply by pattacini** [reply]
`yarpmotorgui` displays the name of the joints.

At any rate, the order is `0:pitch`, `1:roll`, `2:yaw` as per:
- https://github.com/robotology/robots-configuration/blob/master/iCubGenova11/hardware/mechanicals/right_leg-eb8-j0_3-mec.xml#L10

**Reply by jimingda** [reply]
Inside the model.urdf of iCubGazeboV2_5_visuomanip, i found that the yaw angle of r_hip_pitch is -3.14159265359, why not assign the value 0 but -3.14159265359, i mean is there any difference between 0 degree and -360 degree?
```
  <joint name="r_hip_pitch" type="revolute">
    <origin xyz="0.0064515 0.026 -0.119913" rpy="0 0 -3.14159265359"/>
    <axis xyz="-2.22044604925e-16 -1.0 0.0"/>
    <parent link="root_link"/>
    <child link="r_hip_1"/>
    <limit effort="55.5" lower="-0.785398163397" upper="2.33874119767" velocity="5.1"/>
    <dynamics damping="1.0" friction="0.0"/>
  </joint>
```

**Reply by pattacini** [reply]
$`-\pi`$ rad is not $`-360`$ deg but $`-180`$ deg.

**Reply by jimingda** [reply]
Hi @pattacini  
what shall i do if i want to switch to `iCubGazeboV2_7`, whether in docker or in the default ubuntu
And there is `/usr/local/src/robot/robotology-superbuild/build/install/share/iCub/robots/iCubGazeboV2_7/` in the system, but can't find it in gazebo
![2023-08-17_21-38](https://github.com/robotology/community/assets/7019692/e21dc698-c0e2-4daf-a836-b5baf309896a)

**Reply by pattacini** [reply]
Hi @jimingda 

Given that we found recently something to be fixed with `iCubGazeboV2_7` (see https://github.com/robotology/icub-models-generator/issues/245), I would suggest that you stick to `iCubGazeboV2_5` and/or its visuomanip version for the time being.

That said, there's no much difference between `iCubGazeboV2_7` and `iCubGazeboV2_5` for your purposes.

**Comment by xEnVrE** [comment]
> I think that the referenced [robot model](https://github.com/vvv-school/tutorial_cartesian-interface/blob/master/gazebo/worlds/tutorial_cartesian-interface.sdf#L21) isn't shipped with controllable legs. Not sure though.

Indeed, the legs are disabled for that model.
